### PR TITLE
fix(gpws): light only illuminates one frame

### LINKED ISSUE #5554: GPWS light only illuminates for one frame
### Aircraft Version

Development

### Build info

```json
It's been like this for a long time!
```

### Describe the bug

Many of the GPWS callouts trigger the GPWS light above the EFIS Control Panel to illuminate. This illumination only lasts for a single frame at the moment.

**I'll fix this tomorrow.**

### Expected behavior

The GPWS light remains on for the duration of the callout(s).

### Steps to reproduce

1. Fly.
2. Throw yourself at the ground.
3. Observe the GPWS light very briefly illuminating.

### References (optional)

_No response_

### Additional info (optional)

_No response_

### Discord Username (optional)

SjotgunSjonnie#9623

### PR DESCRIPTION
**Pull request by davidwalschots**
Fixes #5554

## Summary of Changes
Many of the GPWS mode types trigger the GPWS light above the EFIS Control Panel to illuminate. This illumination only lasted for a single frame. This PR ensures the light remains illuminated until the mode type becomes inactive.

### Technical notes
`shouldGPWSLightActive` was recalculated every frame, whereas setting `shouldGPWSLightActive` to `true` would only happen during the one frame where a mode type became active. To make fixing this easier, I deduplicated a lot of code in `GPWSComputeLightsAndCallouts` and keep what sounds to play/light to illuminate in a lookup table.

## Screenshots
image

## Additional context

Discord username (if different from GitHub): SjotgunSjonnie#9623

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Review by 2hwk — APPROVED** [review]

**Comment by Samueleonard** [comment]
Quality Assurance Tester Report

Discord: 1samueleonard1
Object of testing: (#5560)
Tier of Testing: (1)
Date: (10/8/2021)

Testing Process: 
. loaded in
. almost flew into a mountain - gpws alerted + light illuminated

Negatives: n/a
Testing Result: (Passed)

Conclusions: it worked + I enjoyed flying into a mountain
Media: n/a

**Comment by NUTTTTTTTT** [comment]
Quality Assurance Tester Report

Discord: █▀█ █▄█ ▀█▀#2123
Object of testing: #PR 5560
Tier of Testing: 1
Date: 11/08/2021

Testing Process:
- Loaded in on runway at EGCC, took off and did the A32NX limbo (how low can you go).
- When terrain/GPWS warning alert raised - GPWS light illuminated, stays illuminated until warning clears.

Negatives: N/A
Testing Results: Passed

Conclusions: Worked well, ended up in a building. Great stuff!
Media:
image

**Review by bouveng — DISMISSED** [review]
Nice refactor.🍸

File: flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_GPWS.js [hunk 1]
```diff
@@ -164,58 +204,27 @@ class A32NX_GPWS {
...
-        if (this.Mode1Code !== this.PrevMode1Code) {
-            if (this.PrevMode1Code === 1) {
-                this.core.soundManager.removePeriodicSound(soundList.sink_rate);
-            }
-            if (this.Mode1Code === 1) {
-                shouldGPWSLightActive |= true;
-                this.core.soundManager.addPeriodicSound(soundList.sink_rate, 1.1);
-            } else if (this.Mode1Code === 2) {
-                shouldGPWSLightActive |= true;
+        this.modes.forEach((mode) => {
```
[the remark this answers is not in the document]
Comment (davidwalschots) [reply flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_GPWS.js:207]: The eslint configuration contains the following rule:

```js
'arrow-parens': ['error', 'always'],
```

File: flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_GPWS.js [hunk 2]
```diff
@@ -225,50 +234,45 @@ class A32NX_GPWS {
...
-        SimVar.SetSimVarValue("L:A32NX_GPWS_Warning_Active", "Bool", shouldGPWSLightActive);
-
-        if (this.Mode5Code !== this.PrevMode5Code) {
-            SimVar.SetSimVarValue("L:A32NX_GPWS_GS_Warning_Active", "Bool", Math.min(this.Mode5Code, 1));
-            if (this.Mode5Code === 1) {
-                // no callout for that (yet)
-            }
-            this.PrevMode5Code = this.Mode5Code;
-        }
+        const illuminateGpwsLight = activeTypes.some((type) => type.gpwsLight);
```
[the remark this answers is not in the document]
Comment (davidwalschots) [reply flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_GPWS.js:237]: The eslint configuration contains the following rule:

```js
'arrow-parens': ['error', 'always'],
```

File: flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_GPWS.js [hunk 3]
```diff
@@ -164,58 +204,27 @@ class A32NX_GPWS {
...
-            }
-            this.PrevMode4Code = this.Mode4Code;
-        }
+            mode.previous = mode.current;
+        });
 
-        const shouldPullUpPlay = this.Mode1Code === 2 || this.Mode2Code === 2;
+        const activeTypes = this.modes.map((mode) => mode.type[mode.current]);
+
+        const shouldPullUpPlay = activeTypes.some((type) => type.pullUp);
```
[the remark this answers is not in the document]
Comment (davidwalschots) [reply flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_GPWS.js:227]: The eslint configuration contains the following rule:

```js
'arrow-parens': ['error', 'always'],
```

File: flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_GPWS.js [hunk 4]
```diff
@@ -164,58 +204,27 @@ class A32NX_GPWS {
...
-                this.core.soundManager.addPeriodicSound(soundList.too_low_terrain, 1.1);
-                shouldGPWSLightActive |= true;
-            }
-            this.PrevMode4Code = this.Mode4Code;
-        }
+            mode.previous = mode.current;
+        });
 
-        const shouldPullUpPlay = this.Mode1Code === 2 || this.Mode2Code === 2;
+        const activeTypes = this.modes.map((mode) => mode.type[mode.current]);
```
[the remark this answers is not in the document]
Comment (davidwalschots) [reply flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_GPWS.js:225]: The eslint configuration contains the following rule:

```js
'arrow-parens': ['error', 'always'],
```

**Pull request merged**